polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -29,78 +31,109 @@ import scala.tools.nsc.interactive.Global
-          _       <- runPredef(kernel, language)
-        } yield kernel
+        case None => IO.pure(Stream.empty)
       }
-    } {
-      _ => launchingKernel.release
     }
+
+  protected def getInterpreter(language: String): Option[LanguageInterpreter[IO]] = Option(interpreters.get(language))
+
+  protected def getOrLaunchInterpreter(language: String): IO[(LanguageInterpreter[IO], Stream[IO, Result])] =
+    Option(interpreters.get(language))
+      .map(interp => IO.pure(interp -> Stream.empty))
+      .getOrElse {
+        launchingInterpreter.acquire.bracket { _ =>
+          Option(interpreters.get(language)).map(interp => IO.pure(interp -> Stream.empty)).getOrElse {
+            for {
+              factory <- IO.fromEither(Either.fromOption(subKernels.get(language), new RuntimeException(s"No interpreter for language $language")))
+              interp  <- taskQueue.runTask(s"Kernel$$$language", s"Starting $language interpreter")(_ => factory.apply(dependencies.getOrElse(language, Nil), symbolTable))
+              _        = interpreters.put(language, interp)
+              results  <- runPredef(interp, language)
+            } yield (interp, results)
+          }
+        } {
+          _ => launchingInterpreter.release
+        }
   }
 
-  protected def withKernel[A](cellId: String)(fn: (Notebook, NotebookCell, LanguageKernel[IO]) => A): IO[A] = for {
+  /**
+    * Perform a task upon a notebook cell, if the interpreter for that cell has already been launched. If the interpreter
+    * hasn't been launched, the action won't be performed and the result will be None.
+    */
+  protected def withLaunchedInterpreter[A](cellId: String)(fn: (Notebook, NotebookCell, LanguageInterpreter[IO]) => IO[A]): IO[Option[A]] = for {
+    notebook <- getNotebook()
+    cell     <- IO(notebook.cell(cellId))
+    interp    = getInterpreter(cell.language)
+    result   <- interp.fold(IO.pure(Option.empty[A]))(interp => fn(notebook, cell, interp).map(result => Option(result)))
+  } yield result
+
+  /**
+    * Perform a task upon a notebook cell, launching the interpreter if necessary. Since this may result in the interpreter's
+    * predef code being executed, and that might produce results, the caller must be prepared to handle the result stream
+    * (which is passed into the given task)
+    */
+  protected def withInterpreter[A](cellId: String)(fn: (Notebook, NotebookCell, LanguageInterpreter[IO], Stream[IO, Result]) => IO[A]): IO[A] = for {
     notebook <- getNotebook()
-    cell     <- IO.fromEither(Either.fromOption(notebook.cells.find(_.id == cellId), new NoSuchElementException(s"Cell $cellId does not exist")))
-    kernel   <- getKernel(cell.language)
-  } yield fn(notebook, cell, kernel)
+    cell     <- IO(notebook.cell(cellId))
+    launch   <- getOrLaunchInterpreter(cell.language)
+    (interp, predefResults) = launch
+    result   <- fn(notebook, cell, interp, predefResults)
+  } yield result
```

Comment:
+1, I like the distinction between `withInterpreter` and `withLaunchedInterpreter`